tsfc/coffee_mode.py

```diff
@@ -0,0 +1,282 @@
+from __future__ import absolute_import, print_function, division
+
+import numpy
+import itertools
+from functools import partial
+from six import iteritems, iterkeys
+from six.moves import filter
+from collections import OrderedDict
+from gem.optimise import (replace_division, make_sum, make_product,
+                          unroll_indexsum, replace_delta, remove_componenttensors)
+from gem.refactorise import Monomial, ATOMIC, COMPOUND, OTHER, collect_monomials
+from gem.node import traversal
+from gem.gem import (Product, Sum, Comparison, Conditional, Division, Indexed,
+                     IndexSum, MathFunction, Power, Failure, one, index_sum,
+                     Terminal, ListTensor, FlexiblyIndexed, LogicalAnd,
+                     LogicalNot, LogicalOr)
+from gem.utils import groupby
+
+
+import tsfc.vanilla as vanilla
+
+flatten = vanilla.flatten
+
+finalise_options = dict(replace_delta=False, remove_componenttensors=False)
+
+
+def Integrals(expressions, quadrature_multiindex, argument_multiindices, parameters):
+    """Constructs an integral representation for each GEM integrand
+    expression.
+
+    :arg expressions: integrand multiplied with quadrature weight;
+                      multi-root GEM expression DAG
+    :arg quadrature_multiindex: quadrature multiindex (tuple)
+    :arg argument_multiindices: tuple of argument multiindices,
+                                one multiindex for each argument
+    :arg parameters: parameters dictionary
+
+    :returns: list of integral representations
+    """
+    # Unroll
+    max_extent = parameters["unroll_indexsum"]
+    if max_extent:
+        def predicate(index):
+            return index.extent <= max_extent
+        expressions = unroll_indexsum(expressions, predicate=predicate)
+    # Choose GEM expression as the integral representation
+    expressions = [index_sum(e, quadrature_multiindex) for e in expressions]
+    expressions = replace_delta(expressions)
+    expressions = remove_componenttensors(expressions)
+    expressions = replace_division(expressions)
+    return optimise_expressions(expressions, argument_multiindices)
+
+
+def optimise_expressions(expressions, argument_multiindices):
+    """Perform loop optimisations on GEM DAGs
+
+    :arg expressions: list of GEM DAGs
+    :arg argument_multiindices: tuple of argument multiindices,
+                                one multiindex for each argument
+
+    :returns: list of optimised GEM DAGs
+    """
+    # Propagate Failure nodes
+    for n in traversal(expressions):
+        if isinstance(n, Failure):
+            return expressions
```

Comment:
I think at this point `Failure` nodes are already eliminated if they aren't actually used. Though it also shouldn't be very hard to let `Failure` nodes pass through these optimisations.